privacyidea/lib/tokenclass.py

```diff
@@ -470,18 +489,29 @@ def update(self, param, reset_failcount=True):
             raise ParameterError('[ParameterError] You may either specify '
                                  'genkey or otpkey, but not both!', id=344)
 
-        if otpKey is not None:
-            self.token.set_otpkey(otpKey, reset_failcount=reset_failcount)
-        elif genkey == 1:
-            otpKey = self._genOtpKey_()
+        if otpKey is None and genkey == 1:
+            otpKey = self._genOtpKey_(key_size)
 
         # otpKey still None?? - raise the exception
         if otpKey is None and self.hKeyRequired is True:
             otpKey = getParam(param, "otpkey", required)
 
         if otpKey is not None:
+            if self.token.rollout_state == "clientwait":
+                # If we have otpkey and the token is in the enrollment-state
+                # generate the new key
+                server_component = self.token.get_otpkey().getKey()
+                client_component = otpKey
+                otpKey = self.generate_symmetric_key(server_component,
+                                                     client_component)
+                self.token.rollout_state = ""
+                self.token.active = True
             self.add_init_details('otpkey', otpKey)
-            self.set_otpkey(otpKey)
+            self.token.set_otpkey(otpKey, reset_failcount=reset_failcount)
+
+        if two_step_init:
+            # After the key is generated, we set "waiting for the client".
+            self.token.rollout_state = "clientwait"
```

Comment:
Ah, you did not provide a parameter ``otpkey``?
In this case privacyIDEA generated the client component and would generate a client component which is as long as the server component. And the dummy ``generate_symmetric_key`` method complains.

In fact, if the token with the given serial number is in ``rollout_state=clientwait`` we would need to **require** a parameter ``otpkey``.

Or take it the other way round: If ``2stepinit=1`` is given, this should only work for non-existing tokens, since this command issues the first step of the two steps. But this parameter is not to be used in the second step.

So imho issuing ``2stepinit=1`` with a serial number of an existing token should result in an error:

"Can not do so, token already exists".